// @opentelemetry/plugin-http/build/src/enums/AttributeNames.js
"use strict";
/*!
 * Copyright 2019, OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Attributes Names according to [OpenTelemetry attributes specs](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#common-attributes)
 */
var AttributeNames;
(function (AttributeNames) {
    AttributeNames["HTTP_HOST"] = "http.host";
    AttributeNames["COMPONENT"] = "component";
    AttributeNames["HTTP_METHOD"] = "http.method";
    AttributeNames["HTTP_TARGET"] = "http.target";
    AttributeNames["HTTP_ROUTE"] = "http.route";
    AttributeNames["HTTP_URL"] = "http.url";
    AttributeNames["HTTP_STATUS_CODE"] = "http.status_code";
    AttributeNames["HTTP_STATUS_TEXT"] = "http.status_text";
    AttributeNames["HTTP_FLAVOR"] = "http.flavor";
    AttributeNames["NET_PEER_IP"] = "net.peer.ip";
    AttributeNames["NET_PEER_PORT"] = "net.peer.port";
    AttributeNames["NET_PEER_NAME"] = "net.peer.name";
    AttributeNames["NET_HOST_IP"] = "net.host.ip";
    AttributeNames["NET_HOST_PORT"] = "net.host.port";
    AttributeNames["NET_HOST_NAME"] = "net.host.name";
    AttributeNames["NET_TRANSPORT"] = "net.transport";
    AttributeNames["IP_TCP"] = "IP.TCP";
    AttributeNames["IP_UDP"] = "IP.UDP";
    AttributeNames["HTTP_SERVER_NAME"] = "http.server_name";
    AttributeNames["HTTP_CLIENT_IP"] = "http.client_ip";
    // NOT ON OFFICIAL SPEC
    AttributeNames["HTTP_ERROR_NAME"] = "http.error_name";
    AttributeNames["HTTP_ERROR_MESSAGE"] = "http.error_message";
    AttributeNames["HTTP_USER_AGENT"] = "http.user_agent";
})(AttributeNames = exports.AttributeNames || (exports.AttributeNames = {}));
