// @opentelemetry/tracing/build/src/config.js
"use strict";
/*!
 * Copyright 2019, OpenTelemetry Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const core_1 = require("@opentelemetry/core");
const scope_base_1 = require("@opentelemetry/scope-base");
/** Default limit for Message events per span */
exports.DEFAULT_MAX_EVENTS_PER_SPAN = 128;
/** Default limit for Attributes per span */
exports.DEFAULT_MAX_ATTRIBUTES_PER_SPAN = 32;
/** Default limit for Links per span */
exports.DEFAULT_MAX_LINKS_PER_SPAN = 32;
/**
 * Default configuration. For fields with primitive values, any user-provided
 * value will override the corresponding default value. For fields with
 * non-primitive values (like `traceParams`), the user-provided value will be
 * used to extend the default value.
 */
exports.DEFAULT_CONFIG = {
    defaultAttributes: {},
    binaryFormat: new core_1.BinaryTraceContext(),
    httpTextFormat: new core_1.HttpTraceContext(),
    logLevel: core_1.LogLevel.DEBUG,
    sampler: core_1.ALWAYS_SAMPLER,
    scopeManager: new scope_base_1.NoopScopeManager(),
    traceParams: {
        numberOfAttributesPerSpan: exports.DEFAULT_MAX_ATTRIBUTES_PER_SPAN,
        numberOfLinksPerSpan: exports.DEFAULT_MAX_LINKS_PER_SPAN,
        numberOfEventsPerSpan: exports.DEFAULT_MAX_EVENTS_PER_SPAN,
    },
};
